Floats really really floats

richardjgowers opened the pull request:
## Description

This is somewhat motivated on an annoying bug we had where (e.g.) a `FloatQuantity` value of `1.0 * unit.nanometer` was creating a different json to a value of `1 * unit.nanometer`.  So instead FloatQuantity should always cast to float.


## Todos
Notable points that this PR has either accomplished or will accomplish.
  - [ ] TODO 1

## Questions
- [ ] Question1

## Status
- [ ] Ready to go

mattwthompson commented:
Interesting idea, I could see it being useful in plenty of situations.

> creating a different json to a value

presumably this is a case where byte-by-byte equality is important?

mattwthompson commented:
I'd be fine with a change to `FloatQuantity` that immediately casts all ints to floats. That would be a breaking change but as long as it's introduced alongside `IntQuantity` I don't see any reason to keep that wart along for the ride. If people want a quantity to be represented as a float, it's reasonable to store it as such, and if people actually want it as an int they'd be able to do that. I can't imagine a case in which something morphs between the two types in a useful way? Like, somehow it's important that it's represented as an int under some conditions and a float under others.

As far as `IntQuantity` goes, if/once you're happy with it I think I'm just going to ask for more tests/higher coverage. IIRC the `if unit_ is Any:` block is covered by models like

```python
class Foo(models.BaseModel):
    a: types.FloatQuantity  # not defining the units
```

Let me know when you want this looked at or if I can help along the way

richardjgowers committed:
ensure FloatQuantity is always a float

now integer inputs to FloatQuantity fields are cast to float

makes for better json() byte-by-byte comparisons

richardjgowers commented:
@mattwthompson I've just included the floats fix here, so these are always floats even if you put in an int.  I'll work up the Int idea separately

mattwthompson commented on openff/models/tests/test_types.py:
```diff
@@ -36,37 +38,46 @@ class Atom(DefaultModel):
-        assert a.quux == 1
+        assert a.quix == 2.0 * unit.nanometer
+        assert isinstance(a.quix.m, float)
+        assert a.quux == 1.0
```
```suggestion
        assert isinstance(a.quux.m, float)
```
This is defined as `int` in the model, no `FloatQuantity` involved, so I think it should not be converted? This is an existing oversight but it'd be nice to fix it here if I'm not mistaken.

I guess equality of raw numbers case doesn't require the two objects to be of the same type:
```
>>> 1 == 1.0
True
```

kinda makes sense and I should have known this already

richardjgowers replied:
yeah this was me being overzealous, it's untouched as an int, but `1 == 1.0`

mattwthompson commented on openff/models/types.py:
```diff
@@ -56,17 +56,22 @@ def validate_type(cls, val):
                     # some custom behavior could go here
                     assert unit_.dimensionality == val.dimensionality
                     # return through converting to some intended default units (taken from the class)
+                    val = float(val.m) * val.u
```
```suggestion
                    val._magnitude = float(val.m)
```

It'd be quite a bit faster to forcefully convert this:

```python
In [6]: %%timeit
   ...: val._magnitude = float(val.m)
   ...:
   ...:
157 ns ± 0.483 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [7]: %%timeit
   ...: float(val.m) * val.u
   ...:
   ...:
11.6 µs ± 35.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

obviously this could break everything horribly as well. Not sure

mattwthompson commented on openff/models/types.py:
```diff
@@ -56,17 +56,22 @@ def validate_type(cls, val):
                 if isinstance(val, str):
                     # could do custom deserialization here?
-                    return unit.Quantity(val).to(unit_)
+                    val = unit.Quantity(val).to(unit_)
+                    return float(val.m) * val.u
```
It'd be nice to avoid double-creating `Quantity`, but I can't find a way to do that just looking at Pint's source and I sure don't want to pre-process the string. Any ideas? If not, that's fine.

mattwthompson reviewed (changes requested):
A couple of questions and one suggested change, otherwise good to go I think

richardjgowers committed:
Update openff/models/types.py

Co-authored-by: Matt Thompson <matt.thompson@openforcefield.org>

richardjgowers committed:
fixup test

richardjgowers committed:
be a bit sharp with handling Quantities

slight performance gain on not creating twice when casting the magnitude field

richardjgowers commented on openff/models/types.py:
```diff
@@ -56,17 +56,23 @@ def validate_type(cls, val):
                     # could do custom deserialization here?
-                    return unit.Quantity(val).to(unit_)
+                    val = unit.Quantity(val).to(unit_)
+                    val._magnitude = float(val._magnitude)
```
@mattwthompson this is a little sharp, but avoids the double creation of Quantity.  Either way is fine

mattwthompson reviewed (approved):
Thanks!

mattwthompson commented:
@richardjgowers this is in 0.0.5 which should be online shortly
